### PR TITLE
Add BitFlip and Sometimes Emission classes

### PR DESCRIPTION
**Pull request by ThomasColthurst**

### CODE REVIEW

**Review by emilyfertig — APPROVED** [review]
Nice!

File: cxx/emissions/bitflip.hh [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  // No hyperparameters to transition!
+  void transition_hyperparameters() {}
+
+  bool sample_corrupted(const bool& clean, std::mt19937* unused_prng) {
+    return !clean;
+  }
+
+  bool propose_clean(const std::vector<bool>& corrupted,
+                           std::mt19937* unused_prng) {
+    return !corrupted[0];
```
[the remark this answers is not in the document]
Comment (ThomasColthurst) [reply cxx/emissions/bitflip.hh:40]: corrupted is expected to contain all the same values.  Specifically, the contract for the corrupted vector is that it is the output of sample_corrupted over repeated calls with the same clean value.

File: cxx/emissions/sometimes.hh [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    }
+    return clean;
+  }
+
+  SampleType propose_clean(const std::vector<SampleType>& corrupted,
+                           std::mt19937* prng) {
+    // We approximate the maximum likelihood estimate by taking the mode of
+    // corrupted.  The full solution would construct BaseEmissor and
+    // BetaBernoulli instances for each choice of clean and picking the
+    // clean with the lowest combined logp_score().
```
[the remark this answers is not in the document]
Comment (ThomasColthurst) [reply cxx/emissions/sometimes.hh]: Yes!  Fixed.

File: cxx/emissions/sometimes.hh [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+template <typename BaseEmissor, typename SampleType = double>
+class Sometimes : public Emission<SampleType> {
+ public:
+  BetaBernoulli bb;
+  BaseEmissor be;
+
+  Sometimes() : bb(nullptr) {};
+
+  void incorporate(const std::pair<SampleType, SampleType>& x) {
+    ++(this->N);
```
[the remark this answers is not in the document]
Comment (ThomasColthurst) [reply cxx/emissions/sometimes.hh:21]: I really really wish I knew.  I originally had it as ++N; and the compiler complained that no variable named N was defined in scope.

File: cxx/emissions/sometimes.hh [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    // We approximate the maximum likelihood estimate by taking the mode of
+    // corrupted.  The full solution would construct BaseEmissor and
+    // BetaBernoulli instances for each choice of clean and picking the
+    // clean with the lowest combined logp_score().
+    std::unordered_map<SampleType, int> counts;
+    SampleType mode;
+    int max_count = 0;
+    for (const SampleType& c: corrupted) {
+      ++counts[c];
+      if (counts[c] > max_count) {
```
[the remark this answers is not in the document]
Comment (ThomasColthurst) [reply cxx/emissions/sometimes.hh:68]: I considered that, but the code to use std::max_element over a map turns out to be many lines that look ugly to me:

auto max = std::max_element(
  std::begin(counts), std::end(counts),
  [] (const auto &p1, const auto &p2) { return p1.second < p2.second; }
)

return max->first;

**Pull request merged**